[WIP] setAttribute with int/long on OSX & Win

ax3l opened the pull request:
Test that writing attributes with non-fixed size types works.

Seems to crash on some environments, such as:
- OSX in Travis, `long` is `long long`
- MSVC, `long` and `int` are identical

Problem is, that only one of them is in the variant list and it seems automatic detection that they are the same is not performed in `variant` on the two obscure OSes above. We should therefore write a trait, that converts types such as `int` to their current, platform-specific fixed-size type during access, e.g. in `Attributable` where `Attribute(value)` is called.

Or in simple words: on OSX and with MSVC on Windows this did not compile:
```c++
        long l = 32;
        s.setAttribute("long", l);
```

since `Attribute(value)` complains that our `Variant` does not contain `long` (but e.g. only `int` and `long long` although `int` is equivalent to `long`).

[First seen here.](https://github.com/openPMD/openPMD-api/pull/330#discussion_r215350563)

ax3l committed:
Tests: setAttribute with int/long

Test HDF5/ADIOS1 of writing and reading attributes with non-fixed
size types.

ax3l committed:
Python: Container Namespace Ambiguity

Fix access to local `detail` namespace.

anokfireball commented:
I'm not yet fully grapsing the problem.
My assumption was that datatype derivation in `Variant` worked similar to what I did manaully in [```determineDatatype()```](https://github.com/openPMD/openPMD-api/blob/0.4.0-alpha/include/openPMD/Datatype.hpp#L103-L140):
```cpp
using Attribute = Variant< T_dtype_enum, T_one, T_two, /* ... */ >;
T_two two = /* ... */;
Attribute(two).dtype == determineDatatype< T_two >();
```

Which it probably does, but as
```cpp
Attribute(T_one()).dtype == Attribute(T_two()).dtype
```
is apparently true for some platforms and combinations, this might break the logic of correlating datatypes by simply looking up the [enum position](https://github.com/openPMD/openPMD-api/blob/0.4.0-alpha/include/openPMD/auxiliary/Variant.hpp#L58).

Now to the problem at the pillar:
> We should therefore write a trait, that converts types such as int to their current, platform-specific fixed-size type during access

The core of this already exists as run-time behaviour in the form of [```determineDatatype()```](https://github.com/openPMD/openPMD-api/blob/0.4.0-alpha/include/openPMD/Datatype.hpp#L103-L140):
Instead of 
```cpp
template< class T_DTYPES, typename ... T >
class Variant
{
public:
    using resource = variantSrc::variant< T ... >;

    Variant(resource r)
            : dtype{static_cast<T_DTYPES>(r.index())},
              m_data{r}
    { }
} //Variant
```
we would have to explicitly cast the passed resource 
```cpp
template< class T_DTYPES, typename ... T >
class Variant
{
public:
    using resource = variantSrc::variant< T ... >;

    template< typename U >
    Variant(U&& u)
            : dtype{determineDatatype< U >()},
              m_data{resource(static_cast< YOUR_TRAIT< U >::type >(u))}
    { }
} //Variant
```

ax3l commented:
@C0nsultant wrote to you on Gitter, we have a solution

ax3l committed:
setAttribute: Explicit C-String Handling

Explicitly handle C-strings in order to avoid implicit casts.
Makes bindings easier.

anokfireball commented on include/openPMD/backend/Attribute.hpp:
```diff
@@ -61,4 +62,81 @@ using Attribute = auxiliary::Variant< Datatype,
+
+        static_assert(
+            wasFound || value,
+            "Type was not found!"
```
Is this the compilation error one will get when doing something like
```cpp
struct NotAnAttributeDType
{ };

Series s = /* ...  */;
auto attr = NotAnAttributeDType();
s.setAttribute("nonsense", attr);
```
?

ax3l replied:
Yes, e.g. on GCC 6.3 you get (I removed the c-string handling for demonstration):
```
openPMD-api/include/openPMD/backend/Attribute.hpp:97:9:
error: static assertion failed: Type was not found!
         static_assert(
         ^~~~~~~~~~~~~
In file included from openPMD-api/include/openPMD/Iteration.hpp:23:0,
                 from openPMD-api/include/openPMD/openPMD.hpp:27,
                 from openPMD-api/test/SerialIOTest.cpp:8:
openPMD-api/include/openPMD/backend/Attributable.hpp: In instantiation of ‘bool openPMD::Attributable::setAttribute(const string&, const T&) [with T = char [59]; std::__cxx11::string = std::__cxx11::basic_string<char>]’:
openPMD-api/test/SerialIOTest.cpp:1290:80:   required from here
openPMD-api/include/openPMD/backend/Attributable.hpp:234:32: warning: the compiler can assume that the address of ‘value’ will always evaluate to ‘true’ [-Waddress]
         it->second = Attribute(static_cast<FixedSize const&>(value));
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
openPMD-api/include/openPMD/backend/Attributable.hpp:240:66: warning: the compiler can assume that the address of ‘value’ will always evaluate to ‘true’ [-Waddress]
                                    std::make_pair(key, Attribute(static_cast<FixedSize const&>(value))));
```

anokfireball commented on include/openPMD/backend/Attribute.hpp:
```diff
@@ -61,4 +62,81 @@ using Attribute = auxiliary::Variant< Datatype,
+     * Takes a fundamental type in T_Input and converts it to its fixed size equivalent.
+     * T_Input is compared against all supported types in @see Attribute .
+     *
+     * Returns the original type of T_Input is already a fixed size type.
```
Typo? 
> the original type *if* T_Input

anokfireball commented on test/SerialIOTest.cpp:
```diff
@@ -899,6 +899,26 @@ TEST_CASE( "hdf5_dtype_test", "[serial][hdf5]" )
+        s.setAttribute("vecLong", std::vector< long >({2147483646, 2147483647}));
+        s.setAttribute("vecUShort", std::vector< unsigned short >({65534u, 65535u}));
+        s.setAttribute("vecUInt", std::vector< unsigned int >({65533u, 65531u}));
+        s.setAttribute("vecULong", std::vector< unsigned long >({65532u, 65530u}));
```
There already is a testcase in [```CoreTest.cpp```](https://github.com/openPMD/openPMD-api/blob/0.4.0-alpha/test/CoreTest.cpp#L26-L80) that does similar checks. As this is (probably?) not supposed to check the workings in the backend, and rather in ```Attribute```, the tests are better suited there.

ax3l replied:
Ah dang, no... I moved the test to `CoreTest` now where a (non-user typical) `Attribute()` is created directly. I did not fix that yet ([see reason below](https://github.com/openPMD/openPMD-api/pull/333#issuecomment-419088770)). I fixed the user-interface `setAttribute`.

ax3l commented:
thx for the review, addressed all points. Will push the other commits after the test has failed, so I see if the issue is still visible as intended.

anokfireball commented on test/CoreTest.cpp:
```diff
@@ -71,6 +71,56 @@ TEST_CASE( "attribute_dtype_test", "[core]" )
+    // translation from non-fixed size integer types
+    a = Attribute(static_cast< short >(1));
+    REQUIRE((
+        Datatype::INT16 == a.dtype || Datatype::INT32 == a.dtype || Datatype::INT64 == a.dtype
```
[```determineDatatype()```](https://github.com/openPMD/openPMD-api/blob/0.4.0-alpha/include/openPMD/Datatype.hpp#L103-L140) provides the datatype of explicit fixed width you are looking for. Would be a little easier to read and more explicit.
```cpp
a = Attribute(static_cast< short >(1));
REQUIRE(determineDatatype< short >() == a.dtype);
```

ax3l replied:
Sounds legit, but wouldn't we then test `determineDatatype()` only against itself?

Anyway, the real test this adds is about the line above, if it compiles:
`a = Attribute(static_cast< short >(1));`

I just added the `REQUIRE` line as well for consistency with the existing checks.

anokfireball replied:
> wouldn't we then test determineDatatype() only against itself?

No, because `Attribute.datatype` is *not* determined by `determineDatatype()`, but by (`std::`/`mpark::`)[`Variant`](https://github.com/openPMD/openPMD-api/blob/0.4.0-alpha/include/openPMD/auxiliary/Variant.hpp#L58) with a subsequent lookup in the `Datatype` enum.

ax3l commented:
@C0nsultant a note on your proposal [here](https://github.com/openPMD/openPMD-api/pull/333#issuecomment-419026604): I would like to move the trait into the `openPMD::Variant` constructor directly because it's more general. The problem is, that we don't have access to all "allowed" types at this point (aka `Attribute`) since they are set later.

Since this will be a larger refactoring, e.g. unifying `Attribute.hpp` and `Variant.hpp` I would leave this open as a possible refactoring later on. For now, I want to get #330 finished with this PR.

ax3l commented:
@C0nsultant will have to push once more. The test needs to go into `SerialTest` again since there `setAttribute` is tested.

ax3l commented on test/CoreTest.cpp:
```diff
@@ -71,6 +71,36 @@ TEST_CASE( "attribute_dtype_test", "[core]" )
     REQUIRE(Datatype::BOOL == a.dtype);
+
+    /* translation from non-fixed size integer types
+     * @todo see
```
@C0nsultant I will leave this prepared test in here, for #333

ax3l commented on include/openPMD/backend/Attribute.hpp:
```diff
@@ -61,4 +62,81 @@ using Attribute = auxiliary::Variant< Datatype,
+    template<bool wasFound, typename U, typename T_Head, typename ...T>
+    struct FindTypeImpl
+    {
+        static constexpr bool value = std::is_same<
```
hm, on the platforms with trouble, `is_same` will still return false although they are the same... Argh!

ax3l replied:
turns out they are not the same. each of them are distinct types, just some happen to implement the same precision. `int32_t` and others are `typedef` aliases of the funnily named ones (`short`, `int`) but not vice versa and not across the latter.

ax3l commented:
Regarding our underlying issue, this might be interesting reads:
https://github.com/MicrosoftDocs/cpp-docs/issues/411
https://en.cppreference.com/w/cpp/language/types#Integer_types

One could now probably either follow this and go for the fundamental types instead of fixed-size types in all relevant comparisons (e.g. all four `short`, `int`, `long`, `long long` instead of the three `int16_t`, `int32_t`, `int64_t`) - which is actually what we already do for floating point types - or we add our own `is_equivalent_fundamental` trait to the functionality inside this PR.

It's also interesting to see that our variadic templates in `Attribute` change the type in the parameter pack from the fixed size types (which are aliases) to their actual types (e.g. `short`, `int`, `long long` on OSX & Win). This also means there is no `cstdint` alias for `long` there.

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -1468,6 +1496,26 @@ TEST_CASE( "adios1_dtype_test", "[serial][adios1]" )
+        s.setAttribute("vecLong", std::vector< long >({2147483646, 2147483647}));
+        s.setAttribute("vecUShort", std::vector< unsigned short >({65534u, 65535u}));
+        s.setAttribute("vecUInt", std::vector< unsigned int >({65533u, 65531u}));
+        s.setAttribute("vecULong", std::vector< unsigned long >({65532u, 65530u}));
```
read: `getAttribute` check missing

ax3l replied:
`LongLong` missing

ax3l commented:
This PR is mainly solved with #337.

I will re-open what is left in here (increased coverage, maybe unique C-string signature for set-attribute, maybe the scope fix in one of the tests) into separate PRs.

anokfireball commented on include/openPMD/backend/Attributable.hpp:
```diff
@@ -97,16 +97,23 @@ class Attributable
      * @param   value   Value of Attribute stored with the provided key.
      * @return  true if key was already present, false otherwise
+     *
+     * @{
```
Out of curiosity:
I am unfamiliar with this systax. Does this tell doxygen to generate only one doc for the both of these?

ax3l replied:
Yes, exactly. It's doxyen group-ing for exactly that use case :)

anokfireball commented on test/SerialIOTest.cpp:
```diff
@@ -1467,7 +1495,33 @@ TEST_CASE( "adios1_dtype_test", "[serial][adios1]" )
         s.setAttribute("vecLongdouble", std::vector< long double >({0.L, std::numeric_limits<long double>::max()}));
         s.setAttribute("vecString", std::vector< std::string >({"vector", "of", "strings"}));
-        s.setAttribute("bool", static_cast< unsigned char >(true));
+        // s.setAttribute("bool", true);
```
See https://github.com/openPMD/openPMD-api/pull/337#discussion_r217307170 in review comment https://github.com/openPMD/openPMD-api/pull/337#pullrequestreview-153581857 (not uncollapsed when using direct link).

anokfireball commented on test/SerialIOTest.cpp:
```diff
@@ -1500,7 +1554,25 @@ TEST_CASE( "adios1_dtype_test", "[serial][adios1]" )
 #endif
     REQUIRE(s.getAttribute("vecString").get< std::vector< std::string > >() == std::vector< std::string >({"vector", "of", "strings"}));
-    REQUIRE(s.getAttribute("bool").get< unsigned char >() == static_cast< unsigned char >(true));
+    // REQUIRE(s.getAttribute("bool").get< bool >() == true);
```
See https://github.com/openPMD/openPMD-api/pull/337#discussion_r217302014 in review comment https://github.com/openPMD/openPMD-api/pull/337#pullrequestreview-153581857 (not uncollapsed when using direct link).

anokfireball reviewed (approved):
LGTM

ax3l commented:
@C0nsultant I will rip this PR apart, since it's otherwise too confusing (https://github.com/openPMD/openPMD-api/pull/333#issuecomment-420705235)
